### DISCUSSION TITLE
Question: Delete leftover XMP metadata when replacing images

### DISCUSSION THREAD

**Discussion by mnmtz** [pikepdf/pikepdf#474] [Q&A] [answered]
After successfully replacing images, the (old) references to XMP metadata object as well as the metadata objects still remain in the PDF file. How can I delete these objects?

**Comment by jbarlow83** [comment]
You can delete the XMP metadata by deleting the Root.Metadata object. 

This XMP belongs to the PDF not the image, unless the PDF generator made your XMP match the image for your specific file. This is a special case so you have to adjust it yourself. (There could be many images in the PDF with different XMP metadata, which is different from the PDF metadata.

**Reply by mnmtz** [reply]
I have already deleted the Root.Metadata object, but these entries still remain in the pdf even if I remove the image. There are 2 references to each xmp object: 1 in the xref table and 1 in the image:

```
<< /BitsPerComponent 8 /ColorSpace 6 0 R /Filter /DCTDecode /Height 1759 /Intent /RelativeColorimetric /Metadata 151 0 R /Name /X /Subtype /Image /Type /XObject /Width 1245 /Length 650729 >>
```

```
151 0 obj
<< /Subtype /XML /Type /Metadata /Length 2344 >>
stream
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="Adobe XMP Core 7.2-c000 79.1b65a79, 2022/06/13-17:46:14        ">
 <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
  <rdf:Description rdf:about=""
    xmlns:xmp="http://ns.adobe.com/xap/1.0/"
    xmlns:dc="http://purl.org/dc/elements/1.1/"
```

**Reply by jbarlow83** [reply]
Image metadata can also be deleted by deleting the Metadata key in the XObject. Some other keys need to be updated as well.
 
Unreferenced objects should be removed/excluded when the PDF is saved.

**Reply by mnmtz** [reply]
I have cleaned the Metadata key in the XObject (image), but remove_unreferenced_resources() still doesn't remove the xmp object - probably because it is referenced in the xref table. How can I delete the reference from the xref table?

**Reply by jbarlow83** [reply, marked as answer]
remove_unreferenced_resources removes objects that are not referenced in the corresponding content stream. (Not your case.)

You need to delete the key from the parent dictionary - deleting the object just deletes that reference.  

It may be the object is referenced from elsewhere in the PDF, so it cannot be removed.

If you inspect the JSON serialization of the PDF you may be able to find what object references the one you are trying to delete.
